File: neanderthal-no-op/src/main/java/au/com/outware/neanderthal/application/NeanderthalApplication.java
package au.com.outware.neanderthal.application;

import android.app.Application;

import java.util.Map;

/**
 * @author timmutton
 */
public class NeanderthalApplication extends Application {
    private Object variant;

    public <T> void initialise(Class<T> klass, Map<String, T> baseVariants, String defaultVariant) {
        variant = baseVariants.get(defaultVariant);
    }

    @SuppressWarnings("unchecked")
    public <T> T getConfiguration() {
        return (T)variant;
    }
}

File: neanderthal/src/main/kotlin/au/com/outware/neanderthal/application/NeanderthalApplication.kt
package au.com.outware.neanderthal.application

import android.app.Application
import au.com.outware.neanderthal.dagger.component.DaggerNeanderthalApplicationComponent
import au.com.outware.neanderthal.dagger.component.NeanderthalApplicationComponent
import au.com.outware.neanderthal.dagger.module.NeanderthallApplicationModule

/**
 * @author timmutton
 */
open class NeanderthalApplication : Application() {
    companion object {
        @JvmStatic
        lateinit var neanderthalApplicationComponent: NeanderthalApplicationComponent
    }

    fun initialise(klass: Class<out Any>, baseVariants: Map<String, Any>, defaultVariant: String) {
        neanderthalApplicationComponent = DaggerNeanderthalApplicationComponent.builder()
                .neanderthallApplicationModule(NeanderthallApplicationModule(this, klass, baseVariants, defaultVariant))
                .build()
        neanderthalApplicationComponent.inject(this)
    }

    @Suppress("UNCHECKED_CAST")
    fun <T> getConfiguration(): T {
        return neanderthalApplicationComponent.getVariantInteractor().getCurrentVariant()?.configuration as T
    }
}